src/server/frontend_wayland/window_wl_surface_role.cpp

```diff
@@ -447,9 +448,10 @@ void mf::WindowWlSurfaceRole::surface_destroyed()
     {
         // "When a client wants to destroy a wl_surface, they must destroy this 'role object' wl_surface"
         // NOTE: the wl_shell_surface specification seems contradictory, so this method is overridden in it's implementation
-        BOOST_THROW_EXCEPTION(std::runtime_error{
-            "wl_surface@" + std::to_string(wl_resource_get_id(surface->resource)) +
-            " destroyed before associated role"});
+        BOOST_THROW_EXCEPTION((wayland::ProtocolError{
+            surface->resource,
+            std::numeric_limits<int32_t>::max(),    // There's no good error code: use a bad one
```

Comment:
I'm not sure how I feel about this, but if we're going to do it this way `wl_subcompositor.cpp` should be changed to throw a protocol error as well.

Reply:
OK